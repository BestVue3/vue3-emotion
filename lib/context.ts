import {
    computed,
    defineComponent,
    PropType,
    inject,
    provide,
    Ref,
    shallowRef,
    watchEffect,
} from 'vue'

import createCache, { EmotionCache } from '@emotion/cache'

function invariant(cond: boolean, message: string): void {
    if (!cond) throw new Error(message)
}

const themeKey = Symbol('emotion-vue-inject-theme-key')
const cacheKey = Symbol('emotion-vue-inject-cache-key')

const defaultThemeRef = shallowRef({})
const defaultCache = createCache({
    key: 'css',
})

const defaultCacheRef = shallowRef(defaultCache)

let cache: EmotionCache = defaultCache

export const ThemeProvider = defineComponent({
    name: 'EmotionVueThemeProvider',
    props: {
        theme: {
            type: Object,
            required: true,
        },
    },
    setup(props, { slots }) {
        const theme = computed(() => props.theme)

        provide(themeKey, theme)

        return () => slots.default && slots.default()
    },
})

export const CacheProvider = defineComponent({
    name: 'EmotionVueThemeProvider',
    props: {
        cache: {
            type: Object as PropType<EmotionCache>,
            required: true,
        },
    },
    setup(props, { slots }) {
        const parentCache = inject(cacheKey, undefined)

        invariant(
            !parentCache,
            `You should never mount two <CacheProvider> in one subtree.`,
        )

        provide(
            cacheKey,
            computed(() => props.cache),
        )

        watchEffect(() => {
            if (props.cache) cache = props.cache
        })

        return () => slots.default && slots.default()
    },
})

export function useTheme<Theme>(): Ref<Theme> {
    return inject(themeKey, defaultThemeRef as Ref<Theme>)
}

export function useCache(): Ref<EmotionCache> {
    return inject(cacheKey, defaultCacheRef)
}
