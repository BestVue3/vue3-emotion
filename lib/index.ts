export * from './emotion-vue'

// export { keyframes } from '@emotion/css'
