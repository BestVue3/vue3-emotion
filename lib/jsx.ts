import { VNodeTypes, VNodeProps, createVNode } from 'vue'

export default function jsx(
    type: VNodeTypes | symbol,
    props?: (Record<string, unknown> & VNodeProps) | null,
    children?: unknown,
    patchFlag?: number,
    dynamicProps?: string[] | null,
    isBlockNode?: boolean,
) {
    if (!props) {
        return createVNode(
            type as any,
            props,
            children,
            patchFlag,
            dynamicProps,
            isBlockNode,
        )
    }

    const { css } = props

    if (!css) {
        return createVNode(
            type as any,
            props,
            children,
            patchFlag,
            dynamicProps,
            isBlockNode,
        )
    }
}
