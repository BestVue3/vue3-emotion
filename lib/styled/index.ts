// @flow
import createStyled from './base'
// import { tags } from './tags'

// bind it to avoid mutating the original function
// const newStyled = styled.bind(null)

// tags.forEach((tagName) => {
//     // $FlowFixMe: we can ignore this because its exposed type is defined by the CreateStyled type
//     newStyled[tagName] = newStyled(tagName)
// })

export default createStyled
