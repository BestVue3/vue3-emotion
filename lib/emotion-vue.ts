// import { CSSInterpolation } from '@emotion/css'
import {
    serializeStyles,
    SerializedStyles,
    Interpolation,
    CSSInterpolation,
} from '@emotion/serialize'
import { EmotionCache, insertStyles } from '@emotion/utils'

// import { tags } from './styled/tags'

import { computed, ComputedRef, Ref, watchEffect } from 'vue'

import { useTheme, useCache } from './context'

// import createStyled from './styled'

function invariant(cond: boolean, message: string): void {
    if (!cond) throw new Error(message)
}

interface StylesObject {
    [name: string]: CSSInterpolation
}

// interface CreateEmotionOptions {
//     themeInjectKey?: symbol | string
//     cacheInjectKey?: symbol | string
// }

// export function css<Theme>(style: CSSInterpolation, theme?: Theme) {
//     const serialized = serializeStyles(
//         [style],
//         cache.registered,
//         theme ? { theme } : undefined,
//     )
//     insertStyles(cache, serialized, false)
//     return `${cache.key}-${serialized.name}`
// }

export function createClasses<P extends StylesObject, Theme>(
    stylesEffect: (theme: Theme) => P,
): ComputedRef<Record<keyof P, string>> {
    const themeRef: Ref<Theme> = useTheme()
    const cacheRef: Ref<EmotionCache> = useCache()

    return computed(() => {
        const theme = themeRef.value
        const cache = cacheRef.value

        const styles = stylesEffect(themeRef.value)

        return Object.keys(styles).reduce((classesMap, name: keyof P) => {
            invariant(
                typeof styles[name] !== 'function',
                `You should not use function on ${name},
                        if you want to get theme you can just get in from 'createClasses' hook
                    `,
            )

            const serialized = serializeStyles(
                [styles[name]],
                cache.registered,
                theme ? { theme } : undefined,
            )
            insertStyles(cache, serialized, false)
            classesMap[name] = `${cache.key}-${serialized.name}`
            return classesMap
        }, {} as Record<keyof P, string>)
    })
}

// function createStaticClasses<P extends StylesObject>(
//     styles: P,
// ): Record<keyof P, string> {
//     return Object.keys(styles).reduce((classesMap, name: keyof P) => {
//         classesMap[name] = css(styles[name])
//         return classesMap
//     }, {} as Record<keyof P, string>)
// }

function insertWithoutScoping(
    cache: EmotionCache,
    serialized: SerializedStyles,
) {
    if (cache.inserted[serialized.name] === undefined) {
        return cache.insert('', serialized, cache.sheet, true)
    }
}

// export const keyframes = (...args: CSSInterpolation[]) => {
//     const serialized = serializeStyles(args, cache.registered)
//     const animation = `animation-${serialized.name}`
//     insertWithoutScoping(cache, {
//         name: serialized.name,
//         styles: `@keyframes ${animation}{${serialized.styles}}`,
//     })

//     return animation
// }
// const injectGlobal = (...args: CSSInterpolation[]) => {
//     const serialized = serializeStyles(args, cache.registered)
//     insertWithoutScoping(cache, serialized)
// }

export function injectGlobalStyle<Theme>(
    stylesEffect: (theme: Theme) => CSSInterpolation | CSSInterpolation[],
) {
    const themeRef: Ref<Theme> = useTheme()
    const cacheRef = useCache()

    watchEffect(() => {
        const styles = stylesEffect(themeRef.value)
        const serialized = serializeStyles(
            Array.isArray(styles) ? styles : [styles],
            cacheRef.value.registered,
        )
        insertWithoutScoping(cacheRef.value, serialized)
    })
}

// const styled = createStyled<Theme>(useCache, useTheme).bind(null)

// tags.forEach((tagName) => {
//     styled[tagName] = styled(tagName)
// })

// return {
//     css,
//     createClasses,
//     createStaticClasses,
//     styled,
//     ThemeProvider,
//     CacheProvider,
//     useCache,
//     useTheme,
//     keyframes,
//     injectGlobal,
//     injectGlobalStyle,
// }

// const defaultEmotion = createEmotion<Record<string, unknown>>()

// // export const css = defaultEmotion.css
export const css = (
    ...args: Interpolation<Record<string, never>>[]
): SerializedStyles => serializeStyles(args, {})
// export const keyframes = defaultEmotion.keyframes
// export const createClasses = defaultEmotion.createClasses
// export const createStaticClasses = defaultEmotion.createStaticClasses
// export const ThemeProvider = defaultEmotion.ThemeProvider
// export const CacheProvider = defaultEmotion.CacheProvider
// export const useCache = defaultEmotion.useCache
// export const useTheme = defaultEmotion.useTheme
// export const styled = defaultEmotion.styled

type Keyframes = {
    name: string
    styles: string
    anim: 1
    toString: () => string
}

export const keyframes = (
    ...args: Interpolation<Record<string, never>>[]
): Keyframes => {
    const insertable = css(...args)
    const name = `animation-${insertable.name}`
    return {
        name,
        styles: `@keyframes ${name}{${insertable.styles}}`,
        anim: 1,
        toString() {
            return `_EMO_${this.name}_${this.styles}_EMO_`
        },
    }
}
