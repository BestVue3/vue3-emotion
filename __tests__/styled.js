import { mount } from '@vue/test-utils'
import { styled, css, ThemeProvider, keyframes, createClasses } from '../lib'

/**
 * [!important]:
 *
 * You must keep your eyes on the reasult of Snapshot
 * if you did not find any styles outputed
 * maybe your demo style is duplicated from others
 *
 * This is because we run test in js-dom
 * and emotion auto cache inserted styles
 * and we have to clear `head` and `body` before every test
 *
 */

styled.h1 = styled('h1')

function matchEmotionSnapshot(wrapper) {
    const elements = Array.from(
        document.querySelectorAll('style[data-emotion]'),
    )

    const str = `
${elements.map((el) => el.innerHTML).join('\n')}

${wrapper.html()}
`

    expect(str).toMatchSnapshot()
}

describe('styled', () => {
    beforeEach(() => {
        // clean up
        document.head.innerHTML = ''
        document.body.innerHTML = ''
    })

    let flex = 0 // to prevent insert same class

    test('no dynamic', () => {
        const H1 = styled.h1`
            float: left;
        `

        matchEmotionSnapshot(mount(() => <H1>hello world</H1>))
    })

    test('basic render', () => {
        const fontSize = 20
        const H1 = styled.h1`
            color: blue;
            font-size: ${fontSize + 'px'};
            @media (min-width: 420px) {
                color: blue;
            }
        `

        matchEmotionSnapshot(mount(() => <H1>hello world</H1>))
    })

    test('basic render with object as style', () => {
        const fontSize = 20
        const H1 = styled.h1({ fontSize })

        matchEmotionSnapshot(mount(() => <H1>hello world</H1>))
    })

    test('object as style', () => {
        const H1 = styled.h1(
            (props) => ({
                fontSize: props.fontSize,
            }),
            (props) => ({ flex: props.flex }),
            { display: 'flex' },
        )

        matchEmotionSnapshot(
            mount(() => (
                <H1 flex={flex++} fontSize={20}>
                    hello world
                </H1>
            )),
        )
    })

    // TODO: won't implemenet this since vue handle `class` differently
    // test('object as className', () => {
    //     const myclass = { toString: () => 'myclass' }
    //     const Comp = styled('div')``

    //     matchEmotionSnapshot(mount(() => <Comp class={myclass} />))
    // })

    test('glamorous style api & composition', () => {
        const H1 = styled.h1((props) => ({ fontSize: props.fontSize }))
        const H2 = styled(H1)((props) => ({ flex: props.flex }), {
            display: 'flex',
        })

        matchEmotionSnapshot(
            mount(() => (
                <H2 flex={flex++} fontSize={20}>
                    hello world
                </H2>
            )),
        )
    })

    test('inline function return value is a function', () => {
        const fontSize = () => 20
        const Blue = styled('h1')`
            font-size: ${() => fontSize}px;
        `

        matchEmotionSnapshot(mount(() => <Blue>hello world</Blue>))
    })

    test('call expression', () => {
        const fontSize = 21
        const Div = styled('div')`
            font-size: ${fontSize}px;
        `

        matchEmotionSnapshot(mount(() => <Div>hello world</Div>))
    })

    test('nested', () => {
        const fontSize = '20px'
        const H1 = styled.h1`
            font-size: ${fontSize};
        `

        const Thing = styled('div')`
            display: flex;
            & div {
                color: green;

                & span {
                    color: red;
                }
            }
        `

        matchEmotionSnapshot(
            mount(() => (
                <Thing>
                    hello <H1>This will be green</H1> world
                </Thing>
            )),
        )
    })

    test('random expressions undefined return', () => {
        const H1 = styled('h1')`
            ${(props) =>
                props.prop &&
                css`
                    font-size: 1rem;
                `};
            color: green;
        `

        matchEmotionSnapshot(
            mount(() => <H1 className={'legacy__class'}>hello world</H1>),
        )
    })

    test('random object expression', () => {
        const margin = (t, r, b, l) => {
            return () => ({
                marginTop: t,
                marginRight: r,
                marginBottom: b,
                marginLeft: l,
            })
        }
        const H1 = styled.h1`
            background-color: hotpink;
            ${(props) => props.prop && { fontSize: '1rem' }};
            ${margin(0, 'auto', 0, 'auto')};
            color: green;
        `

        matchEmotionSnapshot(
            mount(() => (
                <H1 className={'legacy__class'} prop>
                    hello world
                </H1>
            )),
        )
    })

    test('composition', () => {
        const fontSize = 20
        const H1 = styled('h1')`
            font-size: ${fontSize + 'px'};
        `

        const H2 = styled(H1)`
            font-size: ${(fontSize * 2) / 3 + 'px'};
        `

        matchEmotionSnapshot(
            mount(() => <H2 className={'legacy__class'}>hello world</H2>),
        )
    })

    test('input placeholder', () => {
        const Input = styled('input')`
            ::placeholder {
                background-color: green;
            }
        `
        matchEmotionSnapshot(mount(() => <Input>hello world</Input>))
    })

    test('input placeholder object', () => {
        const Input = styled('input')({
            '::placeholder': {
                backgroundColor: 'green',
            },
        })

        matchEmotionSnapshot(mount(() => <Input>hello world</Input>))
    })

    test('object composition', () => {
        const imageStyles = css({ width: 96, height: 96 })

        css([{ color: 'blue' }])

        const red = css([{ color: 'red' }])

        const blue = css([red, { color: 'blue' }])

        const prettyStyles = css([
            {
                borderRadius: '50%',
                transition: 'transform 400ms ease-in-out',
                ':hover': {
                    transform: 'scale(1.2)',
                },
            },
            { border: '3px solid currentColor' },
        ])

        const Avatar = styled('img')`
            ${prettyStyles};
            ${imageStyles};
            ${blue};
        `

        matchEmotionSnapshot(mount(Avatar))
    })

    test('handles more than 10 dynamic properties', () => {
        const H1 = styled('h1')`
            text-decoration: ${'underline'};
            border-right: solid blue 54px;
            background: ${'white'};
            color: ${'black'};
            display: ${'block'};
            border-radius: ${'3px'};
            padding: ${'25px'};
            width: ${'500px'};
            z-index: ${100};
            font-size: ${'18px'};
            text-align: ${'center'};
            border-left: ${(p) => p.theme.blue};
        `

        matchEmotionSnapshot(
            mount(() => (
                <H1 className={'legacy__class'} theme={{ blue: 'blue' }}>
                    hello world
                </H1>
            )),
        )
    })

    test('function in expression', () => {
        const fontSize = 20
        const H1 = styled('h1', { label: 'H1' })`
            font-size: ${fontSize + 'px'};
        `

        const H2 = styled(H1)`
            font-size: ${({ scale }) => fontSize * scale + 'px'};
        `

        matchEmotionSnapshot(
            mount(() => (
                <H2 class={'legacy__class'} scale={2}>
                    hello world
                </H2>
            )),
        )
    })

    test('composition', () => {
        const fontSize = '20px'

        const cssA = css`
            color: blue;
        `

        const cssB = css`
            ${cssA};
            color: red;
        `

        const BlueH1 = styled('h1')`
            ${cssB};
            color: blue;
            font-size: ${fontSize};
        `

        const FinalH2 = styled(BlueH1)`
            font-size: 32px;
        `

        matchEmotionSnapshot(
            mount(() => (
                <FinalH2 class={'legacy__class'} scale={2}>
                    hello world
                </FinalH2>
            )),
        )
    })

    test('higher order component', () => {
        const fontSize = 20
        const Content = styled('div')`
            font-size: ${fontSize}px;
        `

        const squirtleBlueBackground = css`
            background-color: #7fc8d6;
        `

        const flexColumn = (Component) => {
            const NewComponent = styled(Component)`
                ${squirtleBlueBackground};
                background-color: '#343a40';
                flex-direction: column;
            `

            return NewComponent
        }

        const ColumnContent = flexColumn(Content)

        matchEmotionSnapshot(mount(() => <ColumnContent />))
    })

    test('composition based on props', () => {
        const cssA = css`
            color: blue;
        `

        const cssB = css`
            color: green;
        `

        const H1 = styled('h1')`
            ${(props) => (props.a ? cssA : cssB)};
        `

        matchEmotionSnapshot(mount(() => <H1 a>hello world</H1>))

        matchEmotionSnapshot(mount(() => <H1>hello world</H1>))
    })

    test('composition of nested pseudo selectors', () => {
        const defaultLinkStyles = {
            '&:hover': {
                color: 'blue',
                '&:active': {
                    color: 'red',
                },
            },
        }

        const buttonStyles = () => ({
            ...defaultLinkStyles,
            fontSize: '2rem',
            padding: 16,
        })

        const Button = styled('button')(buttonStyles)

        matchEmotionSnapshot(
            mount({
                setup() {
                    const classesRef = createClasses(() => ({
                        button: {
                            '&:hover': {
                                color: 'pink',
                                '&:active': {
                                    color: 'purple',
                                },
                                '&.some-class': {
                                    color: 'yellow',
                                },
                            },
                        },
                    }))
                    return () => (
                        <Button class={classesRef.value.button}>
                            Should be purple
                        </Button>
                    )
                },
            }),
        )
    })

    test('objects', () => {
        const H1 = styled('h1')({ padding: 10 }, (props) => ({
            display: props.display,
        }))
        matchEmotionSnapshot(mount(() => <H1 display="flex">hello world</H1>))
    })

    test('objects with spread properties', () => {
        const defaultText = { fontSize: 20 }
        const Figure = styled('figure')({ ...defaultText })
        matchEmotionSnapshot(mount(() => <Figure>hello world</Figure>))
    })

    test('composing components', () => {
        const Button = styled('button')`
            color: green;
        `
        const OtherButton = styled(Button)`
            display: none;
        `

        const AnotherButton = styled(OtherButton)`
            display: flex;
            justify-content: center;
        `
        matchEmotionSnapshot(
            mount(() => <AnotherButton>hello world</AnotherButton>),
        )
    })

    // TODO: needed in vue3?
    // test('with higher order component that hoists statics', () => {
    //     const superImportantValue = 'hotpink'
    //     const hoc = (BaseComponent) => {
    //         const NewComponent = (props) => (
    //             <BaseComponent someProp={superImportantValue} {...props} />
    //         )
    //         return hoistNonReactStatics(NewComponent, BaseComponent)
    //     }
    //     const SomeComponent = hoc(styled.div`
    //         display: flex;
    //         color: ${(props) => props.someProp};
    //     `)
    //     const FinalComponent = styled(SomeComponent)`
    //         padding: 8px;
    //     `
    //     const tree = renderer.create(<FinalComponent />).toJSON()
    //     expect(tree).toMatchSnapshot()
    // })

    test('prop filtering', () => {
        const Link = styled('a')`
            color: greenfiltering;
        `
        const rest = { m: [3], pt: [4] }

        matchEmotionSnapshot(
            mount(() => (
                <Link
                    a
                    aria-label="some label"
                    b
                    cool
                    data-wow="value"
                    filtering
                    href="link"
                    is
                    prop
                    wow
                    {...rest}
                >
                    hello world
                </Link>
            )),
        )
    })

    test('no prop filtering on non string tags', () => {
        const Link = styled((props) => <a {...props} />)`
            color: green;
        `

        matchEmotionSnapshot(
            mount(() => (
                <Link
                    a
                    aria-label="some label"
                    b
                    cool
                    data-wow="value"
                    filtering
                    href="link"
                    is
                    prop
                    wow
                >
                    hello world
                </Link>
            )),
        )
    })

    test('no prop filtering on string tags started with upper case', () => {
        const Link = styled('SomeCustomLink')`
            color: green;
        `

        matchEmotionSnapshot(
            mount(() => (
                <Link
                    a
                    aria-label="some label"
                    b
                    cool
                    data-wow="value"
                    filtering
                    href="link"
                    is
                    prop
                    wow
                >
                    hello world
                </Link>
            )),
        )
    })

    test('prop filtering on composed styled components that are string tags', () => {
        const BaseLink = styled('a')`
            background-color: hotpink;
        `
        const Link = styled(BaseLink)`
            color: green;
        `

        matchEmotionSnapshot(
            mount(() => (
                <Link
                    a
                    aria-label="some label"
                    b
                    cool
                    data-wow="value"
                    filtering
                    href="link"
                    is
                    prop
                    wow
                >
                    hello world
                </Link>
            )),
        )
    })

    test('throws if undefined is passed as the component', () => {
        expect(
            () =>
                // $FlowFixMe
                styled(undefined)`
                    display: flex;
                `,
        ).toThrowErrorMatchingSnapshot()
    })
    test('withComponent will replace tags but keep styling classes', () => {
        const Title = styled('h1')`
            color: black;
        `
        const Subtitle = Title.withComponent('h2')

        matchEmotionSnapshot(
            mount(() => (
                <article>
                    <Title>My Title</Title>
                    <Subtitle>My Subtitle</Subtitle>
                </article>
            )),
        )
    })
    test('withComponent with function interpolation', () => {
        const Title = styled('h1')`
            color: ${(props) => props.color || 'yellow'};
        `
        const Subtitle = Title.withComponent('h2')

        matchEmotionSnapshot(
            mount(() => (
                <article>
                    <Title>My Title</Title>
                    <Subtitle>My Subtitle</Subtitle>
                </article>
            )),
        )
    })

    test('withComponent does carry styles from flattened component', () => {
        const SomeComponent = styled('div')`
            color: green;
        `
        const AnotherComponent = styled(SomeComponent)`
            color: hotpink;
        `
        const OneMoreComponent = AnotherComponent.withComponent('p')
        matchEmotionSnapshot(mount(() => <OneMoreComponent />))
    })

    test('theming', () => {
        const Div = styled('div')`
            color: ${(props) => props.theme.primary};
        `
        matchEmotionSnapshot(
            mount(() => (
                <ThemeProvider theme={{ primary: 'yellowpick' }}>
                    <Div>this should be yellowpick</Div>
                </ThemeProvider>
            )),
        )
    })
    test('same component rendered multiple times', () => {
        const SomeComponent = styled('div')`
            color: greenpark;
        `

        matchEmotionSnapshot(mount(SomeComponent))

        expect(mount(SomeComponent).html()).toEqual(mount(SomeComponent).html())
        matchEmotionSnapshot(
            mount(() => (
                <SomeComponent>
                    <SomeComponent />
                    <SomeComponent />
                </SomeComponent>
            )),
        )
    })
    test('component selectors', () => {
        let Target = styled('div', {
            // if anyone is looking this
            // please don't do this.
            // use the babel plugin/macro.
            target: 'e322f2d3tbrgf2e0',
        })`
            color: hotpinkse1;
        `
        let SomeComponent = styled('div')`
            color: greengrey;
            ${Target.toString()} {
                color: yellow;
            }
        `
        matchEmotionSnapshot(
            mount(() => (
                <SomeComponent>
                    <Target />
                </SomeComponent>
            )),
        )
    })
    test('keyframes with css call', () => {
        let SomeComp = styled('div')(
            css`
                animation: ${keyframes({
                    'from,to': { color: 'green' },
                    '50%': { color: 'hotpink' },
                })};
            `,
        )
        matchEmotionSnapshot(mount(SomeComp))
    })
})
