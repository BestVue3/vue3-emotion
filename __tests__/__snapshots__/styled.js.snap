// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styled basic render 1`] = `
"
.css-cy3wtx{color:blue;font-size:20px;}
@media (min-width: 420px){.css-cy3wtx{color:blue;}}

<h1 class=\\"css-cy3wtx\\">hello world</h1>
"
`;

exports[`styled basic render with object as style 1`] = `
"
.css-bk9fzy{font-size:20px;}

<h1 class=\\"css-bk9fzy\\">hello world</h1>
"
`;

exports[`styled call expression 1`] = `
"
.css-1ycmxg5{font-size:21px;}

<div class=\\"css-1ycmxg5\\">hello world</div>
"
`;

exports[`styled component selectors 1`] = `
"
.css-1endiyq{color:greengrey;}
.css-1endiyq .e322f2d3tbrgf2e0{color:yellow;}
.css-14exbk5{color:hotpinkse1;}

<div class=\\"css-1endiyq\\"><div class=\\"css-14exbk5 e322f2d3tbrgf2e0\\"></div></div>
"
`;

exports[`styled composing components 1`] = `
"
.css-1lvrex6{color:green;display:none;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-box-pack:center;-ms-flex-pack:center;-webkit-justify-content:center;justify-content:center;}

<button class=\\"css-1lvrex6\\">hello world</button>
"
`;

exports[`styled composition 1`] = `
"
.css-kebyc{font-size:20px;font-size:13.333333333333334px;}

<h1 class=\\"css-kebyc\\">hello world</h1>
"
`;

exports[`styled composition 2`] = `
"
.css-vua55s{color:blue;color:red;color:blue;font-size:20px;font-size:32px;}

<h1 class=\\"legacy__class css-vua55s\\" scale=\\"2\\">hello world</h1>
"
`;

exports[`styled composition based on props 1`] = `
"
.css-1grt4fh{color:blue;}

<h1 class=\\"css-1grt4fh\\">hello world</h1>
"
`;

exports[`styled composition based on props 2`] = `
"
.css-1grt4fh{color:blue;}
.css-11vzs44{color:green;}

<h1 class=\\"css-11vzs44\\">hello world</h1>
"
`;

exports[`styled composition of nested pseudo selectors 1`] = `
"
.css-rl47rh{}
.css-rl47rh:hover{color:pink;}
.css-rl47rh:hover:active{color:purple;}
.css-rl47rh:hover.some-class{color:yellow;}
.css-c1kq0y{font-size:2rem;padding:16px;}
.css-c1kq0y:hover{color:blue;}
.css-c1kq0y:hover:active{color:red;}
.css-c1kq0y:hover{color:pink;}
.css-c1kq0y:hover:active{color:purple;}
.css-c1kq0y:hover.some-class{color:yellow;}

<button class=\\"css-c1kq0y\\">Should be purple</button>
"
`;

exports[`styled function in expression 1`] = `
"
.css-15zrik0-H1{font-size:20px;font-size:40px;}

<h1 class=\\"legacy__class css-15zrik0-H1\\" scale=\\"2\\">hello world</h1>
"
`;

exports[`styled glamorous style api & composition 1`] = `
"
.css-1yodimz{font-size:20px;-webkit-flex:1;-ms-flex:1;flex:1;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}

<h1 fontsize=\\"20\\" class=\\"css-1yodimz\\">hello world</h1>
"
`;

exports[`styled handles more than 10 dynamic properties 1`] = `
"
.css-1mn53vu{-webkit-text-decoration:underline;text-decoration:underline;border-right:solid blue 54px;background:white;color:black;display:block;border-radius:3px;padding:25px;width:500px;z-index:100;font-size:18px;text-align:center;border-left:blue;}

<h1 class=\\"css-1mn53vu\\">hello world</h1>
"
`;

exports[`styled higher order component 1`] = `
"
.css-itdwi{font-size:20px;background-color:#7fc8d6;background-color:'#343a40';-webkit-flex-direction:column;-ms-flex-direction:column;flex-direction:column;}

<div class=\\"css-itdwi\\"></div>
"
`;

exports[`styled inline function return value is a function 1`] = `
"
.css-1u4dn5o{font-size:20px;}

<h1 class=\\"css-1u4dn5o\\">hello world</h1>
"
`;

exports[`styled input placeholder 1`] = `
"
.css-120m7hs{}
.css-120m7hs::-webkit-input-placeholder{background-color:green;}
.css-120m7hs::-moz-placeholder{background-color:green;}
.css-120m7hs:-ms-input-placeholder{background-color:green;}
.css-120m7hs::placeholder{background-color:green;}

<input class=\\"css-120m7hs\\">
"
`;

exports[`styled input placeholder object 1`] = `
"
.css-135lywr{}
.css-135lywr::-webkit-input-placeholder{background-color:green;}
.css-135lywr::-moz-placeholder{background-color:green;}
.css-135lywr:-ms-input-placeholder{background-color:green;}
.css-135lywr::placeholder{background-color:green;}

<input class=\\"css-135lywr\\">
"
`;

exports[`styled keyframes with css call 1`] = `
"
@-webkit-keyframes animation-ocj8pk{from,to{color:green;}50%{color:hotpink;}}
@keyframes animation-ocj8pk{from,to{color:green;}50%{color:hotpink;}}
.css-1y5469o{-webkit-animation:animation-ocj8pk;animation:animation-ocj8pk;}

<div class=\\"css-1y5469o\\"></div>
"
`;

exports[`styled nested 1`] = `
"
.css-ma4a7u{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}
.css-ma4a7u div{color:green;}
.css-ma4a7u div span{color:red;}

<div class=\\"css-ma4a7u\\">hello <h1 class=\\"css-1u4dn5o\\">This will be green</h1> world</div>
"
`;

exports[`styled no dynamic 1`] = `
"
.css-1teb557{float:left;}

<h1 class=\\"css-1teb557\\">hello world</h1>
"
`;

exports[`styled no prop filtering on non string tags 1`] = `
"
.css-zfn0vm{color:green;}

<a a=\\"true\\" aria-label=\\"some label\\" b=\\"true\\" cool=\\"true\\" data-wow=\\"value\\" filtering=\\"true\\" href=\\"link\\" is=\\"true\\" prop=\\"true\\" wow=\\"true\\" class=\\"css-zfn0vm\\"></a>
"
`;

exports[`styled no prop filtering on string tags started with upper case 1`] = `
"


<somecustomlink a=\\"true\\" aria-label=\\"some label\\" b=\\"true\\" cool=\\"true\\" data-wow=\\"value\\" filtering=\\"true\\" href=\\"link\\" is=\\"true\\" prop=\\"true\\" wow=\\"true\\" class=\\"css-zfn0vm\\">hello world</somecustomlink>
"
`;

exports[`styled object as style 1`] = `
"
.css-1nsnxu6{font-size:20px;-webkit-flex:0;-ms-flex:0;flex:0;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}

<h1 fontsize=\\"20\\" class=\\"css-1nsnxu6\\">hello world</h1>
"
`;

exports[`styled object composition 1`] = `
"
.css-18b5fmf{border-radius:50%;-webkit-transition:-webkit-transform 400ms ease-in-out;transition:transform 400ms ease-in-out;border:3px solid currentColor;width:96px;height:96px;color:red;color:blue;}
.css-18b5fmf:hover{-webkit-transform:scale(1.2);-moz-transform:scale(1.2);-ms-transform:scale(1.2);transform:scale(1.2);}

<img class=\\"css-18b5fmf\\">
"
`;

exports[`styled objects 1`] = `
"
.css-1b8zhb0{padding:10px;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}

<h1 display=\\"flex\\" class=\\"css-1b8zhb0\\">hello world</h1>
"
`;

exports[`styled objects with spread properties 1`] = `
"


<figure class=\\"css-bk9fzy\\">hello world</figure>
"
`;

exports[`styled prop filtering 1`] = `
"
.css-gvqk7x{color:greenfiltering;}

<a aria-label=\\"some label\\" data-wow=\\"value\\" href=\\"link\\" is=\\"true\\" class=\\"css-gvqk7x\\">hello world</a>
"
`;

exports[`styled prop filtering on composed styled components that are string tags 1`] = `
"
.css-18aq1cd{background-color:hotpink;color:green;}

<a aria-label=\\"some label\\" data-wow=\\"value\\" href=\\"link\\" is=\\"true\\" class=\\"css-18aq1cd\\">hello world</a>
"
`;

exports[`styled random expressions undefined return 1`] = `
"
.css-14p0lg6{color:green;}

<h1 class=\\"css-14p0lg6\\">hello world</h1>
"
`;

exports[`styled random object expression 1`] = `
"
.css-1cqel1g{background-color:hotpink;font-size:1rem;margin-top:0;margin-right:auto;margin-bottom:0;margin-left:auto;color:green;}

<h1 class=\\"css-1cqel1g\\">hello world</h1>
"
`;

exports[`styled same component rendered multiple times 1`] = `
"
.css-13o9nax{color:greenpark;}

<div class=\\"css-13o9nax\\"></div>
"
`;

exports[`styled same component rendered multiple times 2`] = `
"
.css-13o9nax{color:greenpark;}

<div class=\\"css-13o9nax\\"><div class=\\"css-13o9nax\\"></div><div class=\\"css-13o9nax\\"></div></div>
"
`;

exports[`styled theming 1`] = `
"
.css-1d3h3h2{color:yellowpick;}

<div class=\\"css-1d3h3h2\\">this should be yellowpick</div>
"
`;

exports[`styled throws if undefined is passed as the component 1`] = `
"You are trying to create a styled element with an undefined component.
You may have forgotten to import it."
`;

exports[`styled withComponent does carry styles from flattened component 1`] = `
"
.css-1a2upri{color:green;color:hotpink;}

<p class=\\"css-1a2upri\\"></p>
"
`;

exports[`styled withComponent will replace tags but keep styling classes 1`] = `
"
.css-9xuyn7{color:black;}

<article><h1 class=\\"css-9xuyn7\\">My Title</h1><h2 class=\\"css-9xuyn7\\">My Subtitle</h2></article>
"
`;

exports[`styled withComponent with function interpolation 1`] = `
"
.css-fhhtk3{color:yellow;}

<article><h1 class=\\"css-fhhtk3\\">My Title</h1><h2 class=\\"css-fhhtk3\\">My Subtitle</h2></article>
"
`;
